test/e2e/user_management_test.go

```diff
@@ -156,26 +161,30 @@ func (s *userManagementTestSuite) TestUserDeactivation() {
 		userSignupMember1, murMember1 := s.createAndCheckUserSignup(true, "usertoautodeactivate", "usertoautodeactivate@redhat.com", s.memberAwait, ApprovedByAdmin()...)
 		deactivationExcludedUserSignupMember1, excludedMurMember1 := s.createAndCheckUserSignup(true, "userdeactivationexcluded", "userdeactivationexcluded@excluded.com", s.memberAwait, ApprovedByAdmin()...)
 
-		// Get the base tier that has deactivation enabled
+		// Get the provisioned account's tier
 		baseTier, err := s.hostAwait.WaitForNSTemplateTier("base")
 		require.NoError(t, err)
 
 		// We cannot wait days for testing deactivation so for the purposes of the e2e tests we use a hack to change the provisioned time
 		// to a time far enough in the past to trigger auto deactivation. Subtracting the given period from the current time and setting this as the provisioned
 		// time should test the behaviour of the deactivation controller reconciliation.
-		tierDeactivationDuration := time.Duration(baseTier.Spec.DeactivationTimeoutDays) * time.Hour * 24
+		tierDeactivationDuration := time.Duration(baseTier.Spec.DeactivationTimeoutDays + 1) * time.Hour * 24
 		murMember1, err = s.hostAwait.UpdateMasterUserRecordStatus(murMember1.Name, func(mur *v1alpha1.MasterUserRecord) {
 			mur.Status.ProvisionedTime = &metav1.Time{Time: time.Now().Add(-tierDeactivationDuration)}
 		})
```

Comment:
Maybe stupid question, but shouldn't we also set the `state.deactivating = true` & and add the "Deactivating" condition to the status part in the `UserSignup` resource?

Reply:
Alexey is correct in his summary.  And we do check the deactivating state here already, it's included in the `Deactivated()` condition set:

```
		userSignupMember1, err = s.hostAwait.WaitForUserSignup(userSignupMember1.Name,
			wait.UntilUserSignupHasConditions(Deactivated()...),
			wait.UntilUserSignupHasStateLabel(v1alpha1.UserSignupStateLabelValueDeactivated))
```

Reply:
> As far as I can see this test verifies that the signup with auto-deactivation enabled is auto-deactivated if it reached the timeout.

Yes, but right now the timeout is 3 days after notification was sent, which is something we don't verify here.

> state.deactivating is an internal state. It's implementation detail in this context IMO. We should not rely on this state to test auto-deactivation.

Yes and no, as I said before - with this change we start counting the timeout from the time when we sent the notification and not based on the deactivation timeout set in the config.
Ideally, we should also verify the use-case when we already reach the timeout set in configuration, but we haven't sent the notification yet - in that case, it should not deactivate/remove the user, but it should wait another 3 days. That's actually the reason why we did so many changes in the controllers. Unfortunately, such a test is not here nor in host-operator repo as unit-test (it would be good to add it).

Anyway, my main point was that to make the test working we need to set the `DeactivatingNotificationDays(-1)` - to negative number. TBH, this is really weird, and it's not really clear why we are doing that - the fact how we count when to deactivate the user (and not from which time) is exactly the implementation detail we should avoid here.

Reply:
@sbryzak @alexeykazakov thoughts?

Reply:
We should have tests for it I agree. But they should be additional tests. IMO the test which verifies the expired account should not relay on manual setting the deactivating state. Setting the state to "deactivated" should be enough. And this is what this test does as far as I can see. And changing the default configuration in e2e tests is IMO fine. We do it all over the place. If it's not clear why we change the configuration then we need to comment it in the code.

For proper testing the "deactivating" state we should add separate tests which do care about that intermediate state. If testing it as e2e is tricky then we should make sure that it's fully covered by Unit tests at least (it should be covered by Unit tests regardless).

Reply:
> IMO the test which verifies the expired account should not rely on manual setting the deactivating state

I would rather let the controller set the condition and here just update the transition time than - in the same way as we do with the ProvisionedTime.

> Setting the state to "deactivated" should be enough. And this is what this test does as far as I can see. 

Not really - these tests don't verify manual deactivation but automatic, see for example this one: https://github.com/sbryzak/toolchain-e2e/blob/12370e3c64e78e3b0249abe134bd353c51ad4dc4/test/e2e/user_management_test.go#L157-L217
By mistake, I added the comment to the test with the name: `"tests for tiers with automatic deactivation disabled"` I should have added it to the one that I referenced `"tests for tiers with automatic deactivation enabled"` to make it more clear. Anyway, in both cases, we should rely on the step of setting the `deactivating` state and not on setting `DeactivatingNotificationDays` to a negative value to skip this step.

> And changing the default configuration in e2e tests is IMO fine.  We do it all over the place.

yes, but we set meaningful values, not negative ones to skip some state in the UserSignup lifecycle. 

> For proper testing the "deactivating" state we should add separate tests which do care about that intermediate state. 

This is exactly the place where it should be. The whole flow is not verified anywhere - it should be verified in these tests where we deal with the automatic approval.

Reply:
Initially @sbryzak added the deacitivating state in all tests which needed a deactivated account. And I asked not to do that.
So, sorry I didn't get the context of this particular test into account when replying.
I agree that in this particular test where we are testing the automated deactivation (not manual) it's better to test the entire flow. But I would test it differently:
- Provision an account.
- Update the provisioning timestamp to something old enough
- Wait for the deactivating state/status
- Update the deactivating notification status timestamp to something old enough
- Wait for the account to be deactivated

So, we do not modify any default configuration, don't skip any steps of the flow and don't set any states manually. We would test exactly what we want: automated flow.

Reply:
This flow is exactly what I had in my mind - sorry for not being clear :+1: 
@sbryzak could you please take care of it?